python/src/iceberg/table/partitioning.py

```diff
@@ -64,3 +67,100 @@ def __str__(self):
 
     def __repr__(self):
         return f"PartitionField(field_id={self.field_id}, name={self.name}, transform={repr(self.transform)}, source_id={self.source_id})"
+
+    def __hash__(self):
+        return hash((self.source_id, self.field_id, self.name, self.transform))
+
+
+class PartitionSpec:
+    """
+    PartitionSpec capture the transformation from table data to partition values
+
+    Attributes:
+        schema(Schema): the schema of data table
+        spec_id(int): any change to PartitionSpec will produce a new specId
+        fields(List[PartitionField): list of partition fields to produce partition values
+        last_assigned_field_id(int): auto-increment partition field id starting from PARTITION_DATA_ID_START
+    """
+
+    PARTITION_DATA_ID_START: int = 1000
+
+    def __init__(self, schema: Schema, spec_id: int, fields: Tuple[PartitionField], last_assigned_field_id: int):
+        self._schema = schema
+        self._spec_id = spec_id
+        self._fields = fields
+        self._last_assigned_field_id = last_assigned_field_id
+        # derived
+        self.fields_by_source_id: Dict[int, List[PartitionField]] = {}
+
+    @property
+    def schema(self) -> Schema:
+        return self._schema
+
+    @property
+    def spec_id(self) -> int:
+        return self._spec_id
+
+    @property
+    def fields(self) -> Tuple[PartitionField]:
+        return self._fields
+
+    @property
+    def last_assigned_field_id(self) -> int:
+        return self._last_assigned_field_id
+
+    def __eq__(self, other):
+        return self.spec_id == other.spec_id and self.fields == other.fields
+
+    def __str__(self):
+        if self.is_unpartitioned():
```

Comment:
Rather than special casing `is_unpartitioned`, I think this should just construct the inner string differently. The Java implementation doesn't use special cases.

Reply:
I think it's a bit nasty to construct it exactly like what Java implementation does, so the special casing is essentially simplified from https://github.com/apache/iceberg/blob/master/api/src/main/java/org/apache/iceberg/PartitionSpec.java#L299. The unpartitioned differ from rest with 1 less `\n` right before the `]`

so here's my alternative way to construct, not sure how do you like it compare to what I have right now.



```python
    def __str__(self):
        result_str = "["
        for partition_field in self.fields:
            result_str += f"\n  {str(partition_field)}"
        if self.is_unpartitioned():
            result_str += "]"
        else:
            result_str += "\n]"
        return result_str
```

Python `string.join(Iterable)` does not have a way to attach head and tail, so I figured it might be easier to see it this way as in `f"{head}{delimiter.join(partition_fields_in_str)}{tail}"`.

But I can certainly change quickly if above is what you prefer

Reply:
> I think it's a bit nasty to construct it exactly like what Java implementation does

I don't think it is necessary to construct it exactly like Java. But it should not have a special case here, since `is_unpartitioned` can include specs with `void` transforms. This should show all of the transforms and not special case.

Reply:
>I don't think it is necessary to construct it exactly like Java. 

got it, thanks for clarify

> is_unpartitioned can include specs with void transforms

But I am a bit behind on this. From what I can tell, the void transform can never be unpartitioned per existing java code https://github.com/apache/iceberg/blob/32841a273b6dda897f6efece4ebb178f8ad54719/api/src/main/java/org/apache/iceberg/PartitionSpec.java#L97-L103. And if there's one void transform added to such PartitionSpec, then I assume there's one partition field with voidTransform as its Transform, which is not `unpartitioned` like below 

```
> PartitionSpec ps = PartitionSpec.builderFor(SCHEMA).alwaysNull("ts").build();
> System.out.println(ps.toString());
[
  1000: ts_null: void(5)
]

> System.out.println(ps.isUnpartitioned());
false
```